contribute/style-guides/e2e-playwright.md

```diff
@@ -0,0 +1,299 @@
+# End-to-end tests
+
+Grafana Labs uses a minimal [homegrown solution](../../e2e/utils/index.ts) built on top of [Playwright](https://playwright.dev/) for its end-to-end (E2E) tests.
+
+Important notes:
+
+- We generally store all element identifiers ([CSS selectors](https://mdn.io/docs/Web/CSS/CSS_Selectors)) within the framework for reuse and maintainability.
+- We generally do not use stubs or mocks as to fully simulate a real user.
+- We also use Playwright for the [plugins' end-to-end tests](contribute/style-guides/e2e-plugins.md).
+
+## Framework structure
+
+Our framework structure is inspired by [Martin Fowler's Page Object](https://martinfowler.com/bliki/PageObject.html).
+
+- **`Selector`**: A unique identifier that is used from the E2E framework to retrieve an element from the browser
+- **`Page`**: An abstraction for an object that contains one or more `Selector` identifiers with the `visit` function to go to the page.
+- **`Component`**: An abstraction for an object that contains one or more `Selector` identifiers but without the `visit` function
+- **`Flow`**: An abstraction that contains a sequence of actions on one or more `Page` abstractions that can be reused and shared between tests
+
+## How to create an end-to-end Playwright test
+
+### Basic example
+
+Let's start with a simple [JSX](https://reactjs.org/docs/introducing-jsx.html) example containing a single input field that we want to populate during our E2E test:
+
+```jsx
+<input className="gf-form-input login-form-input" type="text" />
+```
+
+It is possible to target the field with a CSS selector like `.gf-form-input.login-form-input`. However, doing so is a brittle solution because style changes occur frequently.
+
+Furthermore, there is nothing that signals to future developers that this input is part of an E2E test. At Grafana, we use `data-testid` attributes as our preferred way of defining selectors. See [Aria-Labels vs data-testid](#aria-labels-vs-data-testid) for more details.
+
+```jsx
+<input data-testid="Username input field" className="gf-form-input login-form-input" type="text" />
+```
+
```

Comment:
minor, but can we remove the `gf-form-input` class/selector from this example since it's deprecated

Reply:
Good catch! 🙌